runtime/stdlib.js

```diff
@@ -352,85 +352,129 @@ function caml_compare_val (a, b, total) {
-          if (tb === 250) {
-            b = b[1];
-            continue;
-          } else if (ta != tb) {
-            return (ta < tb)?-1:1;
-          } else {
-            switch (ta) {
-            case 248: {
-              // Object
-              var x = caml_int_compare(a[2], b[2]);
-              if (x != 0) return x;
-              break;
-            }
-            case 251: {
-              caml_invalid_argument("equal: abstract value");
-            }
-            case 255: {
-              // Int64
-              var x = caml_int64_compare(a, b);
-              if (x != 0) return x;
-              break;
-            }
-            default:
-              if (a.length != b.length) return (a.length < b.length)?-1:1;
-              if (a.length > 1) stack.push(a, b, 1);
-            }
-          }
-        } else
-          return 1;
-      } else if (b instanceof MlBytes ||
-                 (b instanceof Array && b[0] === (b[0]|0))) {
-        return -1;
-      } else if (typeof a != "number") {
-        if(a && a.compare) {
-          var cmp = a.compare(b,total);
-          if (cmp != 0) return cmp;
-        }
-        else if(a && a.caml_custom) {
-          if(caml_custom_ops[a.caml_custom].compare) {
-            var cmp = caml_custom_ops[a.caml_custom].compare(a,b,total)
-            if (cmp != 0) return cmp;
+          if(!comp)
+            caml_invalid_argument("compare: abstract value");
+          var x = comp(a,b,total);
+          if (x != 0) return x;
+          break;
+        case 1256: // compare function
+          var x = a.compare(b,total);
+          if(x != x | 0)
+            caml_invalid_argument("compare: abstract value");
+          x = x | 0;
+          if (x != 0) return x;
+          break;
+        case 1000: // Number
+        case 1001: // The rest
+          // Here we can be in the following cases:
+          // - a and b are primitive types, comparison works as expected.
+          // - a or b aren't primitive types, comparison always return false.
+          if (a < b) return -1;
+          if (a > b) return 1;
+          if (a != b) {
+            if (!total) return NaN;
+            if (a == a) return 1;
+            if (b == b) return -1;
           }
-          caml_invalid_argument("compare: abstract value");
-        }
-        else if (typeof a == "function") {
-          caml_invalid_argument("compare: functional value");
-        }
-        else {
-          caml_invalid_argument("compare: abstract value");
-        }
-      } else {
-        if (a < b) return -1;
-        if (a > b) return 1;
-        if (a != b) {
-          if (!total) return NaN;
-          if (a == a) return 1;
-          if (b == b) return -1;
+          break;
+        case 246: // Lazy_tag
+        case 254: // Double_array
+        default: // Block with other tag
+          if (a.length != b.length) return (a.length < b.length)?-1:1;
+          if (a.length > 1) stack.push(a, b, 1);
```

Comment:
Is this not just pushing the same objects back on the stack?

Reply:
The stack looks like `[offset, objectleft, objectright, offset, objectleft, objectright]`.
Here we push the same objects with offset `1`. The next time we pick elements from the stack we'll pick `a[1]` and `b[1]` and push `stack.push(a, b, i + 1);`

```
    var i = stack.pop();
    b = stack.pop();
    a = stack.pop();
    if (i + 1 < a.length) stack.push(a, b, i + 1);
    a = a[i];
    b = b[i];
```